### PR TITLE
Support graph-based approximate nearest neighbors search (cagra on nn_descent/ivf_pq)

### PR DESCRIPTION
**Pull request by lijinf2**

### CODE REVIEW

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/knn.py [hunk 1]
```diff
@@ -916,6 +924,20 @@ class ApproximateNearestNeighbors(
...
         Note this class constructs the kmeans index independently on individual data partition (or maxRecordPerBatch if Arrow is enabled).
 
+        When algorithm is 'cagra':
+            parameters for index construction
+            * build_algo: (str, default = 'ivf_pq') algorithm to build graph index, can be either 'ivf_pq' or 'nn_descent'. nn_descent is expected to be generally faster than ivf_pq.
+            * intermediate_graph_degree: (int, default = 128) an intermediate variable used during graph index construction
+            * graph_degree: (int, default = 64) the degree of each node in the final graph index
+
+            parameters for search, full list in cuvs python API documentation.
+            * itopk_size: (int, default = 64) number of intermediate search results retained during the search. Larger value improves the search accuracy but increases the search time.
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/knn.py:934]: I guess the intermediate topk related params need to be no smaller than the top level one?   Our code should enforce/set this.

File: python/src/spark_rapids_ml/knn.py [hunk 2]
```diff
@@ -916,6 +924,20 @@ class ApproximateNearestNeighbors(
         if set, algoParam is used to configure the algorithm, on each data partition (or maxRecordsPerBatch if Arrow is enabled) of the item_df.
         Note this class constructs the kmeans index independently on individual data partition (or maxRecordPerBatch if Arrow is enabled).
 
+        When algorithm is 'cagra':
+            parameters for index construction
+            * build_algo: (str, default = 'ivf_pq') algorithm to build graph index, can be either 'ivf_pq' or 'nn_descent'. nn_descent is expected to be generally faster than ivf_pq.
+            * intermediate_graph_degree: (int, default = 128) an intermediate variable used during graph index construction
+            * graph_degree: (int, default = 64) the degree of each node in the final graph index
+
+            parameters for search, full list in cuvs python API documentation.
```
[the remark this answers is not in the document]
Comment (lijinf2) [reply python/src/spark_rapids_ml/knn.py]: Yeah, added the link and tested new docs locally.

File: python/src/spark_rapids_ml/knn.py [hunk 3]
```diff
@@ -916,6 +924,20 @@ class ApproximateNearestNeighbors(
         if set, algoParam is used to configure the algorithm, on each data partition (or maxRecordsPerBatch if Arrow is enabled) of the item_df.
         Note this class constructs the kmeans index independently on individual data partition (or maxRecordPerBatch if Arrow is enabled).
 
+        When algorithm is 'cagra':
+            parameters for index construction
+            * build_algo: (str, default = 'ivf_pq') algorithm to build graph index, can be either 'ivf_pq' or 'nn_descent'. nn_descent is expected to be generally faster than ivf_pq.
+            * intermediate_graph_degree: (int, default = 128) an intermediate variable used during graph index construction
+            * graph_degree: (int, default = 64) the degree of each node in the final graph index
```
[the remark this answers is not in the document]
Comment (lijinf2) [reply python/src/spark_rapids_ml/knn.py]: It seems irrelevant that can be set independently, as I can understand. 

graph_degree is the number of neighbors of every node in the final graph index. It is used in the index construction phase to control the size of graph index. 

topk is used in the search phase to control the number of returned neighbors. 

When a query visits a node of the graph index, all of its neighbors will be evaluated to update the current topk results. A larger graph_degree should be helpful for a given k, and the two parameters can be set independently.

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/knn.py (hunk 1 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/knn.py]: Yes, added validation on the itopk_size parameter. cuml increases the value to the nearest multiple of 32 and requires the value to be larger than or equal to k.

File: python/tests/test_approximate_nearest_neighbors.py [hunk 4]
```diff
@@ -451,6 +514,136 @@ def test_ivfpq(
...
+) -> None:
+
+    itopk_size = 64 if "itopk_size" not in algo_params else algo_params["itopk_size"]
+    import math
+
+    internal_topk_size = math.ceil(itopk_size / 32) * 32
+    n_neighbors = 50
+    error_msg = ""
+    if internal_topk_size < n_neighbors:
+        error_msg = f"cuVS increases itopk_size to be closest multiple of 32 and expects the value, i.e. {internal_topk_size}, to be larger than or equal to k, i.e. {n_neighbors}."
```
Comment (eordentlich) [inline python/tests/test_approximate_nearest_neighbors.py:632]: same comment as above.

**Review by eordentlich — COMMENTED** [review]
Looks good.  Minor comment.

File: python/tests/test_approximate_nearest_neighbors.py (hunk 4 above)
Comment (lijinf2) [reply python/tests/test_approximate_nearest_neighbors.py]: done

**Comment by lijinf2** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**